### PR TITLE
Show deprecation status of operators and sensors on the docs operator list page

### PR DESCRIPTION
**Pull request by pankajkoti**
This PR introduces enhancements to the documentation for operators and sensors, specifically to indicate their deprecation status and suggest replacements where applicable.

# Changes

## New Attributes
Added `is_deprecated` and `post_deprecation_replacement` attributes to operator and sensor classes.
1. Deprecation Information: The `is_deprecated` attribute is a boolean that flags whether a class is deprecated.
2. Replacement Suggestion: The `post_deprecation_replacement` attribute provides a string with the import path for the recommended replacement.

## Script changes that builds the operators list
The script `docs/_ext/traverse_operators_sensors.py` that builds in to the documentation the list of operators and sensors now uses the above newly added attributes to denote whether the operators have been deprecated

closes: #1567

### CODE REVIEW

**Comment by pankajkoti** [comment]
Built docs preview

screencapture-localhost-63342-astronomer-providers-docs-build-html-providers-operators-and-sensors-list-html-2024-08-19-23_26_19

File: setup.cfg [hunk 1]
```diff
@@ -139,6 +135,12 @@ all =
     paramiko
     snowflake-sqlalchemy>=1.4.4  # Temporary solution for https://github.com/astronomer/astronomer-providers/issues/958, we should pin apache-airflow-providers-snowflake version after it pins this package to great than or equal to 1.4.4.
 
+docs =
+    sphinx
+    sphinx-autoapi
+    sphinx-copybutton
+    astronomer-providers[all]
```
Comment (kaxil) [inline setup.cfg:142]: Hmm, this is recursive, no?

You just need to install:

```
pip install -e .[docs]
```

**Review by kaxil — APPROVED** [review]
Nice, alternative way if you didn't wanted to change the source code:

```diff
diff --git a/docs/_ext/traverse_operators_sensors.py b/docs/_ext/traverse_operators_sensors.py
index 13891b1..9078036 100644
--- a/docs/_ext/traverse_operators_sensors.py
+++ b/docs/_ext/traverse_operators_sensors.py
@@ -2,6 +2,7 @@ import ast
 import os
 from pathlib import Path
 from typing import List, Tuple
+import re
 
 from docutils import nodes
 from sphinx.util.docutils import SphinxDirective
@@ -22,12 +23,13 @@ def collect_elements(
     directory_path: str,
     files: List[str],
     element_type: str,
-    elements_list: List[Tuple[str, str]],
+    elements_list: List[Tuple[str, str, bool]],
     current_module_path: str,
 ):
     """
     Checks that ``Async`` class definitions exist using ``ast.parse`` in the given files, that those are
-    ``operators/sensors`` and appends the operator/sensor along with its import path to the given output list.
+    ``operators/sensors`` and appends the operator/sensor along with its import path and deprecation status
+    to the given output list.
 
     :param directory_path: path of the directory in which the given ``files`` are located
     :param files: list of files to look for async operator/sensor class definitions
@@ -47,7 +49,35 @@ def collect_elements(
                 if isinstance(element, ast.ClassDef):
                     element_name = element.name
                     if element_type in element_name and ASYNC_SUBSTRING in element_name:
-                        elements_list.append((element_name, module_import_path))
+                        is_deprecated = False
+                        replacement_class = ""
+                        for class_element in element.body:
+                            if isinstance(class_element, ast.FunctionDef) and class_element.name == "__init__":
+                                for stmt in class_element.body:
+                                    if isinstance(stmt, ast.Expr) and isinstance(stmt.value, ast.Call):
+                                        if isinstance(stmt.value.func, ast.Attribute):
+                                            if stmt.value.func.attr == "warn" and stmt.value.func.value.id == "warnings":
+                                                is_deprecated = True
+                                                # Concatenate the warning message
+                                                warning_message = ""
+                                                for arg in stmt.value.args:
+                                                    if isinstance(arg, ast.Str):
+                                                        warning_message += arg.s
+                                                # Extract the replacement class from the concatenated message
+                                                match = re.search(
+                                                    r":class: `~([^`]+)`|`([^`]+)`",
+                                                    warning_message
+                                                )
+                                                if match:
+                                                    replacement_class_path = match.group(1) or match.group(2)
+                                                    # Transform to import statement format
+                                                    if replacement_class_path:
+                                                        parts = replacement_class_path.rsplit('.', 1)
+                                                        if len(parts) == 2:
+                                                            replacement_class = (
+                                                                f"from {parts[0]} import {parts[1]}"
+                                                            )
+                        elements_list.append((element_name, is_deprecated, module_import_path, replacement_class))
 
 
 def search_providers():
@@ -79,15 +109,18 @@ class TraverseOperatorsSensors(SphinxDirective):
         """Generates raw html to list the operators and sensors parsed using `ast`."""
         operators, sensors = search_providers()
         operators_html = (
-            "<h3>Operators</h3> " "<table>" "<th>#</th>" "<th>Operator name</th>" "<th>Import path</th>"
+            "<h3>Operators</h3> " "<table>" "<th>#</th>" "<th>Operator name</th>" "<th>Deprecated </th>" 
+            "<th>Import path</th>" "<th>Replacement import path </th>"
         )
         for index, operator in enumerate(operators, start=1):
-            class_def_link = operator[1].replace(".", "/") + "/index.html#" + operator[1] + "." + operator[0]
+            class_def_link = operator[2].replace(".", "/") + "/index.html#" + operator[2] + "." + operator[0]
             operators_html += (
                 f"<tr>"
                 f"<td>{index}</td>"
                 f"<td><span><a id={operator[0]}>{operator[0]}</a></span></td>"
-                f"<td><span><pre><code class='python'>from {operator[1]} import {operator[0]}</code></pre></span></td>"
+                f"<td><span>{'✅' if operator[1] else '❌' }</span></td>"
+                f"<td><span><pre><code class='python'>from {operator[2]} import {operator[0]}</code></pre></span></td>"
+                f"<td><span><pre><code class='python'>{operator[3]}</code></pre></span></td>"
                 f"</tr>"
             )
             # The below script generates the URL for the class definition by extracting the selected doc version from
@@ -101,14 +134,19 @@ class TraverseOperatorsSensors(SphinxDirective):
             )
         operators_html += "</table> <br/>"
 
-        sensors_html = "<h3>Sensors</h3>" "<table>" "<th>#</th>" "<th>Sensor name</th>" "<th>Import path</th>"
+        sensors_html = (
+            "<h3>Sensors</h3>" "<table>" "<th>#</th>" "<th>Sensor name</th>" "<th>Deprecated </th>" 
+            "<th>Import path</th>" "<th>Replacement import path </th>"
+        )
         for index, sensor in enumerate(sensors, start=1):
-            class_def_link = sensor[1].replace(".", "/") + "/index.html#" + sensor[1] + "." + sensor[0]
+            class_def_link = sensor[2].replace(".", "/") + "/index.html#" + sensor[2] + "." + sensor[0]
             sensors_html += (
                 f"<tr>"
                 f"<td>{index}</td>"
                 f"<td><span><a id={sensor[0]}>{sensor[0]}</a></span></td>"
-                f"<td><span><pre><code class='python'>from {sensor[1]} import {sensor[0]}</code></pre></span></td>"
+                f"<td><span>{'✅' if sensor[1] else '❌' }</span></td>"
+                f"<td><span><pre><code class='python'>from {sensor[2]} import {sensor[0]}</code></pre></span></td>"
+                f"<td><span><pre><code class='python'>{sensor[3]}</code></pre></span></td>"
                 f"</tr>"
             )
             sensors_html += (
```

**Comment by kaxil** [comment]
Alternatively, you can have a single column with old & new path:

image


Git diff for that:

```diff
diff --git a/docs/_ext/traverse_operators_sensors.py b/docs/_ext/traverse_operators_sensors.py
index 13891b1..4304a29 100644
--- a/docs/_ext/traverse_operators_sensors.py
+++ b/docs/_ext/traverse_operators_sensors.py
@@ -2,6 +2,7 @@ import ast
 import os
 from pathlib import Path
 from typing import List, Tuple
+import re
 
 from docutils import nodes
 from sphinx.util.docutils import SphinxDirective
@@ -22,12 +23,13 @@ def collect_elements(
     directory_path: str,
     files: List[str],
     element_type: str,
-    elements_list: List[Tuple[str, str]],
+    elements_list: List[Tuple[str, str, bool]],
     current_module_path: str,
 ):
     """
     Checks that ``Async`` class definitions exist using ``ast.parse`` in the given files, that those are
-    ``operators/sensors`` and appends the operator/sensor along with its import path to the given output list.
+    ``operators/sensors`` and appends the operator/sensor along with its import path and deprecation status
+    to the given output list.
 
     :param directory_path: path of the directory in which the given ``files`` are located
     :param files: list of files to look for async operator/sensor class definitions
@@ -47,7 +49,35 @@ def collect_elements(
                 if isinstance(element, ast.ClassDef):
                     element_name = element.name
                     if element_type in element_name and ASYNC_SUBSTRING in element_name:
-                        elements_list.append((element_name, module_import_path))
+                        is_deprecated = False
+                        replacement_class = ""
+                        for class_element in element.body:
+                            if isinstance(class_element, ast.FunctionDef) and class_element.name == "__init__":
+                                for stmt in class_element.body:
+                                    if isinstance(stmt, ast.Expr) and isinstance(stmt.value, ast.Call):
+                                        if isinstance(stmt.value.func, ast.Attribute):
+                                            if stmt.value.func.attr == "warn" and stmt.value.func.value.id == "warnings":
+                                                is_deprecated = True
+                                                # Concatenate the warning message
+                                                warning_message = ""
+                                                for arg in stmt.value.args:
+                                                    if isinstance(arg, ast.Str):
+                                                        warning_message += arg.s
+                                                # Extract the replacement class from the concatenated message
+                                                match = re.search(
+                                                    r":class: `~([^`]+)`|`([^`]+)`",
+                                                    warning_message
+                                                )
+                                                if match:
+                                                    replacement_class_path = match.group(1) or match.group(2)
+                                                    # Transform to import statement format
+                                                    if replacement_class_path:
+                                                        parts = replacement_class_path.rsplit('.', 1)
+                                                        if len(parts) == 2:
+                                                            replacement_class = (
+                                                                f"from {parts[0]} import {parts[1]}"
+                                                            )
+                        elements_list.append((element_name, is_deprecated, module_import_path, replacement_class))
 
 
 def search_providers():
@@ -79,17 +109,29 @@ class TraverseOperatorsSensors(SphinxDirective):
         """Generates raw html to list the operators and sensors parsed using `ast`."""
         operators, sensors = search_providers()
         operators_html = (
-            "<h3>Operators</h3> " "<table>" "<th>#</th>" "<th>Operator name</th>" "<th>Import path</th>"
+            "<h3>Operators</h3> " "<table>" "<th>#</th>" "<th>Operator name</th>" "<th>Deprecated </th>" 
+            "<th>Import path</th>"
         )
         for index, operator in enumerate(operators, start=1):
-            class_def_link = operator[1].replace(".", "/") + "/index.html#" + operator[1] + "." + operator[0]
-            operators_html += (
-                f"<tr>"
-                f"<td>{index}</td>"
-                f"<td><span><a id={operator[0]}>{operator[0]}</a></span></td>"
-                f"<td><span><pre><code class='python'>from {operator[1]} import {operator[0]}</code></pre></span></td>"
-                f"</tr>"
-            )
+            class_def_link = operator[2].replace(".", "/") + "/index.html#" + operator[2] + "." + operator[0]
+            if operator[1]:
+                operators_html += (
+                    f"<tr>"
+                    f"<td>{index}</td>"
+                    f"<td><span><a id={operator[0]}>{operator[0]}</a></span></td>"
+                    f"<td style='text-align: center;'>✅</td>"
+                    f"<td><b>Old Path:</b>\n <span><pre><code class='python'>from {operator[2]} import {operator[0]}</code></pre></span>\n <b>New Path:</b> \n<span><pre><code class='python'>{operator[3]}</code></pre></span></td>" 
+                    f"</tr>"
+                )
+            else:
+                operators_html += (
+                    f"<tr>"
+                    f"<td>{index}</td>"
+                    f"<td><span><a id={operator[0]}>{operator[0]}</a></span></td>"
+                    f"<td style='text-align: center;'>❌</td>"
+                    f"<td><span><pre><code class='python'>from {operator[2]} import {operator[0]}</code></pre></span></td>"
+                    f"</tr>"
+                )
             # The below script generates the URL for the class definition by extracting the selected doc version from
             # the current browser URL location.
             operators_html += (
@@ -101,16 +143,30 @@ class TraverseOperatorsSensors(SphinxDirective):
             )
         operators_html += "</table> <br/>"
 
-        sensors_html = "<h3>Sensors</h3>" "<table>" "<th>#</th>" "<th>Sensor name</th>" "<th>Import path</th>"
+        sensors_html = (
+            "<h3>Sensors</h3>" "<table>" "<th>#</th>" "<th>Sensor name</th>" "<th>Deprecated </th>" 
+            "<th>Import path</th>"
+        )
         for index, sensor in enumerate(sensors, start=1):
-            class_def_link = sensor[1].replace(".", "/") + "/index.html#" + sensor[1] + "." + sensor[0]
-            sensors_html += (
-                f"<tr>"
-                f"<td>{index}</td>"
-                f"<td><span><a id={sensor[0]}>{sensor[0]}</a></span></td>"
-                f"<td><span><pre><code class='python'>from {sensor[1]} import {sensor[0]}</code></pre></span></td>"
-                f"</tr>"
-            )
+            class_def_link = sensor[2].replace(".", "/") + "/index.html#" + sensor[2] + "." + sensor[0]
+            if sensor[1]:
+                sensors_html += (
+                    f"<tr>"
+                    f"<td>{index}</td>"
+                    f"<td><span><a id={sensor[0]}>{sensor[0]}</a></span></td>"
+                    f"<td style='text-align: center;'>✅</td>"
+                    f"<td><b>Old Path:</b>\n <span><pre><code class='python'>from {sensor[2]} import {sensor[0]}</code></pre></span>\n <b>New Path:</b> \n<span><pre><code class='python'>{sensor[3]}</code></pre></span></td>"
+                    f"</tr>"
+                )
+            else:
+                sensors_html += (
+                    f"<tr>"
+                    f"<td>{index}</td>"
+                    f"<td><span><a id={sensor[0]}>{sensor[0]}</a></span></td>"
+                    f"<td style='text-align: center;'>❌</td>"
+                    f"<td><span><pre><code class='python'>from {sensor[2]} import {sensor[0]}</code></pre></span></td>"
+                    f"</tr>"
+                )
             sensors_html += (
                 f"<script> "
                 f"var base_url = '' + window.location.pathname.split('/providers/')[0] + '/_api/';"
```

**Comment by kaxil** [comment]
The change you have is also sufficient and does not rely on regex which is good :) so don't feel obligated to change anything. I just wanted to show more options

File: setup.cfg (hunk 1 above)
Comment (pankajkoti) [reply setup.cfg:142]: no, it does not cause a recursion as `all` does not include the dependencies for `docs`, but the `docs` now need all the packages to be installed to evaluate the attributes on the classes.

**Comment by pankajkoti** [comment]
> The change you have is also sufficient and does not rely on regex which is good :) so don't feel obligated to change anything. I just wanted to show more options

Thanks Kaxil for the efforts in showing the alternatives. Mine is definitely some fast-food but yours is food cooked with love :)

For now, I have applied the suggestion to improve the visualisation by having a single column for the import path with the old-new path approach and also including the visual ✅ & ❌ for deprecation status. I am also equally inclined to prefer the definite attribute approach compared to the regex searching for warning texts, so kept that logic as it is, hope that's okay 🤞🏽 

Docs preview with the latest commit after applying the suggestion
screencapture-localhost-63342-astronomer-providers-docs-build-html-providers-operators-and-sensors-list-html-2024-08-20-14_57_19

**Comment by pankajkoti** [comment]
I'm merging this PR but please leave a comment if there are more suggestions and I will address them on priority, thanks for the review and suggestions!

**Pull request merged**